### ISSUE TITLE
Playing sounds freezes game

### ISSUE THREAD

**Issue by angband-import** [angband/angband#4037]
**Reported by vext01 on 11 Sep 2016 20:14 UTC**
Hi,

Angband seems to work great on OpenBSD (once working around ncurses detection -- reported as separate ticket). However, when sound is enabled, the game freezes, presumably at the point it needs to play a sound -- usually I am killing a thief in the town as a test.

Here is a trace of the frozen thread (SIGINT is me pressing ctrl+c):

```
Thread 1 received signal SIGINT, Interrupt.
0x000006d6ec539d4a in _thread_sys___thrsleep () at <stdin>:2
2       <stdin>: No such file or directory.
(gdb) bt
#0  0x000006d6ec539d4a in _thread_sys___thrsleep () at <stdin>:2
#1  0x000006d75741925c in _rthread_mutex_lock (mutexp=0x6d7a4314910, trywait=0, abstime=0x0) at /usr/src/lib/librthread/rthread_sync.c:151
#2  0x000006d719f03fde in SDL_mutexP () from /usr/local/lib/libSDL.so.8.0
#3  0x000006d6f739f80a in Mix_FadeInMusicPos () from /usr/local/lib/libSDL_mixer.so.5.0
#4  0x000006d4c5514c8c in game_event_dispatch (type=EVENT_SOUND, data=0x7f7ffffbb910) at game-event.c:43
#5  0x000006d4c5514f24 in event_signal_message (type=3313682592, t=<optimized out>, s=0x0) at game-event.c:180
#6  0x000006d4c5533928 in msgt (type=2, fmt=<optimized out>) at message.c:443
#7  0x000006d4c5565499 in py_attack_real (fear=<optimized out>, x=<optimized out>, y=<optimized out>) at player-attack.c:398
#8  py_attack (y=7, x=33) at player-attack.c:442
#9  0x000006d4c55086f6 in move_player (dir=<optimized out>, disarm=true) at cmd-cave.c:1061
#10 0x000006d4c5508d4d in do_cmd_walk (cmd=<optimized out>) at cmd-cave.c:1247
#11 0x000006d4c5509582 in process_command (ctx=CMD_GAME, cmd=0x6d4c5831520 <cmd_queue+2112>) at cmd-core.c:224
#12 0x000006d4c5509640 in cmdq_pop (c=1466044800) at cmd-core.c:252
#13 0x000006d4c55156ec in process_player () at game-world.c:632
#14 0x000006d4c55164c3 in run_game_loop () at game-world.c:729
#15 0x000006d4c5587175 in play_game (new_game=false) at ui-game.c:429
#16 0x000006d4c55b5df0 in main (argc=1, argv=0x7f7ffffbca38) at main.c:556

```

Sorry I have not had time to look deeper.

OpenBSD-current/amd64 on angband-4.0.5.

Thanks

**Comment by angband-import** [comment]
**Comment by NickMcConnell on 21 May 2017 03:47 UTC**
Marking all OpenBSD bugs as such and setting them to Future

**Comment by angband-import** [comment]
**Comment by vext01 on 7 Apr 2019 11:13 UTC**
Hi,

This is still an issue with angband-4.1.3.

**Comment by backwardsEric** [comment]
This still occurs in OpenBSD 6.9 with SDL 1.2 from pkg_add (sdl-1.2.15p11, sdl-image-1.2.12p4, sdl-ttf-2.0.11p3, sdl-mixer-1.2.12p5).  It doesn't happen with the OpenBSD 6.9 and the SDL 2 front end (SDL2 from pkg_add; sdl2-2.0.11p3, sdl2-image-2.0.5, sdl2-ttf-2.0.15, sdl2-mixer-2.0.4).

**Comment by backwardsEric** [comment]
Given that /usr/local/lib/playmus hangs if asked to play pls_tone_blurk.mp3 from the Angband sounds directory (it also hangs on the other sound files I tried from there) this is a problem with the OpenBSD's libraries (sdl-mixer or its dependency, smpeg, are the prime suspects) that, at best, Angband can work around somehow (reformat the sound files?).

For reproducing the hang with playmus, if $ANGBAND is the path to the Angband's top-level source directory (or the directory that has Angband's lib directory as a subdirectory), then

```
playmus ${ANGBAND}/lib/sounds/pls_tone_blurk.mp3
```

hangs.  Running it under gdb and interrupting it at the point of the hang gives these details:

```
>info threads
2 thread 603235 futex () at /tmp/-:3
*1 thread 247780 futex() at /tmp/-:3
>bt
0 futex () at /tmp/-:3
1 0x000000db57ad17a7 in _rthread_mutex_timedlock () at include/machine/atomic.h:94
2 0x000000da8c5de18c in SDL_mutex/libP () from /usr/local/libSDL.so.8.0
3 0x000000dac14be458 in Mix_PlayingMusic () from /usr/local/lib/libSDL_mixer.so.5.0
4 0x000000d8684cea98 in ?? from /usr/local/bin/playmus
5 0x000000d8684ce271 in ?? from /usr/local/bin/playmus
6 0x0000000000000000 in ?? ()
>thread 2
[Switching to thread 2 (thread 603235)] 0 futex () at /tmp/-:3
3 in /tmp/-
>bt
0 futex () at /tmp/-:3
1 0x000000da6f489320 in sem_wait () at synch.h:41
2 0x000000da8c5dde08 in SDL_SemWait () from /usr/local/lib/libSDL.so.8.0
3 0x000000db59fb1393 in MPEG_ring::NextReadBuffer () from /usr/local/lib/libsmpeg.so.2.0
4 0x000000db59fcc210 in Play_MPEGautdio () from /usr/local/lib/libsmpeg.so.2.0
5 0x000000dac14bcc08 in music_mixer () from /usr/local/lib/libSDL_mixer.so.5.0
6 0x000000dac14ba474 in mix_channels () from /usr/local/lib/libSDL_mixer.so.5.0
7 0x000000da8c59014e in SDL_RunAudio () from /usr/local/lib/libSDL.so.8.0
8 0x000000da8c59c01c in SDL_RunThread () from /usr/local/lib/libSDL.so.8.0
9 0x000000da8c5ddc29 in RunThread () from /usr/local/lib/libSDL.so.8.0
10 0x000000da6f48af51 in _rthread_start () from /usr/src/lib/librthread/rthread.c:96
11 0x000000db57a6eb9a in _tfork_thread () at /usr/src/lib/libc/arch/amd64/sys/tfork_thread.S:84
12 0x000000db57a6eb9a in _tfork_thread () at /usr/src/lib/libc/arch/amd64/sys/tfork_thread.S:84
Previous frame identical to this frame (corrupt stack?)
```

**Comment by backwardsEric** [comment]
A patch to smpeg that fixes this problem was posted to OpenBSD's ports mailing list.  The message with the patch can be found here, https://marc.info/?l=openbsd-ports&m=162187552125402&w=2 .  Given that OpenBSD's smpeg package doesn't have a dedicated maintainer and SDL 1.2, which is using smpeg to handle reading in the MP3 file, isn't under active development, it's not clear when or if this'll be fixed in OpenBSD.

Converting the sound files to stereo (which would avoid the internal monaural to stereo conversion that triggers the hang) or to another format (wav?) likely would be a workaround to resolve the problem from Angband's end. Or one could use the SDL2 mixer (but then it's not possible to use the SDL front end) since that works.

**Comment by dagit** [comment]
I don't know if it's the same bug, but I was just playing on void linux and hit a sound related freeze. I tried to open a locked door and the game froze. I'm not sure if it was the first locked door I tried to open on that character, but it may have been.

angband 4.2.1

I believe these are the versions of the dependencies:

smpeg-0.4.5_5
SDL-1.2.15_13
SDL_mixer-1.2.12_14

**Comment by backwardsEric** [comment]
If it occurs again and you've got a debugger installed and are familiar with attaching a debugger to a running process, could you get a stack trace for what Angband's doing when frozen?  Some bugs that cause crashes have been fixed in versions since 4.2.1, but I don't remember any reports of freezes due to SDL sound on Linux.  On OpenBSD, the issue affected any sound that was played and was related to a patch that OpenBSD had applied to smpeg.

**Comment by dagit** [comment]
I decided to build it myself so I could have all the bug fixes in 4.2.4. I also decided to make it depend on sdl2. I've played a bit since then and no lockup yet. However, I don't know if the one I saw yesterday was repeatable. At any rate, if you don't hear from me again it means no more freezes.

**Comment by backwardsEric** [comment]
That bug, https://github.com/angband/angband/issues/4687 ,  was supposed to have been squashed in 4.2.3 and I can't reproduce it with the current development version (running on Debian 11).  Is there any chance that src/main.c or src/ui-game.c are not from 4.2.4 in what you built (or src/main.o or src/ui-game.o were built from older source code)?

**Comment by dagit** [comment]
I built it from the source release tar gz for 4.2.4 because I didn't want to figure out how to satisfy autoconf. Maybe there's more to that bug that previously thought?

**Comment by backwardsEric** [comment]
It's possible, but could you double-check that /home/dagit/local-install/games/angband reports that it is 4.2.4.  The 'V' command in-game will report the version.

**Comment by dagit** [comment]
It does. It's definitely 4.2.4.

**Comment by dagit** [comment]
That version seems to fix the sound crash on respawn. Thanks.
